### PR TITLE
0.12.4

### PR DESCRIPTION
**Pull request by FlorianRappl**
# New Pull Request

For more information, see the `CONTRIBUTING` guide.

## Prerequisites

Please make sure you can check the following boxes:

- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] All new and existing tests passed

### Contribution Type

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [x] Bug fix (non-breaking change which fixes an issue, please reference the issue id)
- [x] New feature (non-breaking change which adds functionality, make sure to open an associated issue first)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes

### Description

This version represents a slight iteration and improvement on the anyway very stable and so far looking great 0.12 release train. Most likely the final version before 1.0.0.

Notably, it introduces:

- A flexible templating system for scaffolding new Piral instances and for the core pilet setup (reusable across different app shells; our standard templates "empty" and "default" are automatically in there)
- Improved CLI plugin support using things such as `survey`
- Updated the dependencies (most importantly `dets`)
- Added `piral-breadcrumbs` to introduce breadcrumbs very easily
- Added support for Webpack 5 with `piral-cli-webpack5`
- Added support for `packageOverrides` in the metadata

### Remarks

No remarks.

### CODE REVIEW

File: src/plugins/piral-breadcrumbs/src/useBreadcrumbs.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import { useGlobalState } from 'piral-core';
+import { BreadcrumbRegistration } from './types';
+
+function getExact(breadcrumbs: Array<BreadcrumbRegistration>, path: string) {
+  const [bc] = breadcrumbs.filter((m) => m.matcher.test(path));
```
Comment (nmalyschkin) [inline src/plugins/piral-breadcrumbs/src/useBreadcrumbs.ts:5]: [Array.find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) would do the job.


What happens if there are multiple registrations, that match the given path. Wouldn't the registration order dictate the outcome of `getExact`?

File: src/plugins/piral-breadcrumbs/src/useBreadcrumbs.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  if (path) {
+    return getExact(breadcrumbs, path) || getClosest(breadcrumbs, path);
+  }
+
+  return undefined;
+}
+
+export function useBreadcrumbs(path: string) {
+  const breadcrumbs = useGlobalState((s) => s.registry.breadcrumbs);
+  const list = Object.keys(breadcrumbs).map((id) => breadcrumbs[id]);
```
Comment (nmalyschkin) [inline src/plugins/piral-breadcrumbs/src/useBreadcrumbs.ts:35]:
```js
const list = Object.values(breadcrumbs);
```

File: src/plugins/piral-feeds/src/withFeed.tsx [hunk 3]
```diff
@@ -13,7 +13,7 @@ export function withFeed<TData, TItem, TProps>(
     if (!loaded) {
       return <PiralLoadingIndicator />;
     } else if (data) {
-      return <Component data={data} {...props} />;
+      return <Component {...props} data={data} />;
```
Comment (nmalyschkin) [inline src/plugins/piral-feeds/src/withFeed.tsx:16]: What's your take on this syntax:
```jsx
return <Component {...{ ...props, data }} />;
```
I believe this change has been made so data overwrites props. Imho the concatenated syntax makes this more explicit and noticalbe.

**Review by nmalyschkin — APPROVED** [review]

File: src/plugins/piral-breadcrumbs/src/useBreadcrumbs.ts (hunk 2 above)
Comment (FlorianRappl) [reply src/plugins/piral-breadcrumbs/src/useBreadcrumbs.ts:35]: No - https://caniuse.com/object-values

File: src/plugins/piral-feeds/src/withFeed.tsx (hunk 3 above)
Comment (FlorianRappl) [reply src/plugins/piral-feeds/src/withFeed.tsx:16]: I think they are equivalent, but the used version is (at least for me) more transparent. It says "the react component should take all the values from props and set data explicitly", while the other version says "the react component should take all the values from an object that takes all the values from props and sets data to data".

File: src/plugins/piral-breadcrumbs/src/useBreadcrumbs.ts (hunk 1 above)
Comment (FlorianRappl) [reply src/plugins/piral-breadcrumbs/src/useBreadcrumbs.ts:5]: No - https://caniuse.com/array-find (always have an eye for IE11; we should avoid polyfills if they don't bring any advantage).

We only take the first one.

File: src/plugins/piral-breadcrumbs/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+  interface PiralCustomState {}
+
+  interface PiralCustomActions {
+    /**
+     * Registers a new breadcrumb.
+     * @param name The name of the breadcrumb.
+     * @param value The breadcrumb registration.
+     */
+    registerBreadcrumb(name: string, value: BreadcrumbRegistration): void;
```
[the remark this answers is not in the document]
Comment (FlorianRappl) [reply src/plugins/piral-breadcrumbs/src/types.ts:15]: I think we had that discussion already; and I'd say we should offer this API, too. However, I would not introduce it now - I would introduce it for all registerX APIs then.

[the remark this answers is not in the document]
Comment (FlorianRappl) [reply src/plugins/piral-breadcrumbs/src/types.ts:15]: Can you open an issue to track this ("return disposer from register APIs")? Thanks @nmalyschkin !

[the remark this answers is not in the document]
Comment (nmalyschkin) [reply src/plugins/piral-breadcrumbs/src/types.ts:15]: Alright, will do.

File: src/plugins/piral-feeds/src/withFeed.tsx (hunk 3 above)
Comment (nmalyschkin) [reply src/plugins/piral-feeds/src/withFeed.tsx:16]: I read the syntax as: "pass the props and set/overwrite data". This is a question of taste, just wanted to mention it.

**Pull request merged**